File: backend/fretquiz/src/main/java/kdp/FretQuiz/websocket/Response.java
package kdp.FretQuiz.websocket;

import kdp.FretQuiz.game.Game;
import kdp.FretQuiz.game.Guess;
import kdp.FretQuiz.user.User;

import java.util.List;

/**
 * A Response is a message sent from the server to the client.
 * Each Response has a Response.Type and, optionally, some associated data.
 */
public interface Response {
    Type type();

    enum Type {
        FLASH_MESSAGE,
        LOGGED_IN,
        LOGGED_OUT,
        GAME_IDS,
        GAME_CREATED,
        GAME_JOINED,
        GAME_STARTED,
        GUESS_RESULT,
        PLAYER_JOINED,
        GAME_UPDATED,
        ROUND_STARTED;
    }

    record FlashMessage(Type type, String message) implements Response {
        public FlashMessage(String message) {
            this(Type.FLASH_MESSAGE, message);
        }
    }

    record LoggedIn(Type type, User user) implements Response {
        public LoggedIn(User user) {
            this(Type.LOGGED_IN, user);
        }
    }

    record LoggedOut(Type type) implements Response {
        public LoggedOut() {
            this(Type.LOGGED_OUT);
        }
    }

    record GameCreated(Type type, Game game) implements Response {
        public GameCreated(Game game) {
            this(Type.GAME_CREATED, game);
        }
    }

    record GameIds(Type type, List<String> gameIds) implements Response {
        public GameIds(List<String> gameIds) {
            this(Type.GAME_IDS, gameIds);
        }
    }

    record GameJoined(Type type, Game game) implements Response {
        public GameJoined(Game game) {
            this(Type.GAME_JOINED, game);
        }
    }

    record GameStarted(Type type, Game game) implements Response {
        public GameStarted(Game game) {
            this(Type.GAME_STARTED, game);
        }
    }

    record GuessResult(Type type, Guess guess, Game game) implements Response {
        public GuessResult(Guess guess, Game game) {
            this(Type.GUESS_RESULT, guess, game);
        }
    }

    record PlayerJoined(Type type, String message) implements Response {
        public PlayerJoined(String message) {
            this(Type.PLAYER_JOINED, message);
        }
    }

    record GameUpdated(Type type, Game game) implements Response {
        public GameUpdated(Game game) {
            this(Type.GAME_UPDATED, game);
        }
    }

    record RoundStarted(Type type, Game game) implements Response {
        public RoundStarted(Game game) {
            this(Type.ROUND_STARTED, game);
        }
    }
}

File: frontend/fretquiz/src/websocket/response.ts
import { Game, GameInfo, Guess } from '../game/types';
import { User } from "../user/userSlice";

export type ResponseType =
  'FLASH'
  | 'LOGGED_IN'
  | 'LOGGED_OUT'
  | 'GAME_IDS'
  | 'GAME_INFOS'
  | 'GAME_CREATED'
  | 'GAME_JOINED'
  | 'GAME_STARTED'
  | 'GUESS_RESULT'
  | 'PLAYER_JOINED'
  | 'GAME_UPDATED'
  | 'ROUND_STARTED'
  | 'GAME_OVER'
  | 'GAME_COUNTDOWN';

export interface Response {
  type: ResponseType
}

export interface Flash extends Response {
  type: 'FLASH',
  message: string
}

// user messages

export interface LoggedIn extends Response {
  type: 'LOGGED_IN',
  user: User
}

export interface LoggedOut extends Response {
  type: 'LOGGED_OUT'
}

// game messages

export interface GameCreated extends Response {
  type: 'GAME_CREATED',
  game: Game
}

export interface GameIds extends Response {
  type: 'GAME_IDS',
  gameIds: string[]
}

export interface GameInfos extends Response {
  type: 'GAME_INFOS',
  gameInfos: GameInfo[]
}

export interface GameJoined extends Response {
  type: 'GAME_JOINED',
  game: Game
}

export interface GameStarted extends Response {
  type: 'GAME_STARTED',
  game: Game
}

export interface GuessResult extends Response {
  type: 'GUESS_RESULT',
  guess: Guess,
  game: Game
}

export interface PlayerJoined extends Response {
  type: 'PLAYER_JOINED'
  message: string
}

export interface GameUpdated extends Response {
  type: 'GAME_UPDATED',
  game: Game
}

export interface RoundStarted extends Response {
  type: 'ROUND_STARTED',
  game: Game
}

export interface GameOver extends Response {
  type: 'GAME_OVER',
  game: Game
}

export interface GameCountdown extends Response {
  type: 'GAME_COUNTDOWN',
  secondsLeft: number
}
